src/main/java/com/avispl/symphony/dal/device/core510i/QRCCommunicator.java

```diff
@@ -0,0 +1,436 @@
+				this.logger.debug("Sending: " + data + " to: " + this.host + " port: " + this.port);
+			}
+
+			String[] response = this.internalSend(data);
+
+			if (this.logger.isDebugEnabled()) {
+				this.logger.debug("Received response: " + Arrays.toString(response) + " from: " + this.host + " port: " + this.port);
+			}
+
+			if (this.logger.isTraceEnabled()) {
+				this.logger.trace("Received response: " + Arrays.toString(response) + " from: " + this.host + " port: " + this.port);
+			}
+
+			this.status.setLastTimestamp(System.currentTimeMillis());
+			return response;
+		} catch (CommandFailureException ex1) {
+			if (this.logger.isErrorEnabled()) {
+				this.logger.error("Command failed " + data + " to: " + this.host + " port: " + this.port + " connection state: " + this.status.getConnectionState(), ex1);
+			}
+
+			this.status.setLastTimestamp(System.currentTimeMillis());
+			throw ex1;
+		} catch (SocketTimeoutException ex2) {
+			if (this.logger.isDebugEnabled()) {
+				this.logger.debug("Time out while sending command: " + data + " to: " + this.host + " port: " + this.port + " connection state: " + this.status.getConnectionState() + " error: ", ex2);
+			}
+
+			this.status.setLastError(ex2);
+			this.status.setConnectionState(ConnectionState.Unknown);
+			this.destroyChannel();
+
+			if (retryOnError) {
+				return this.send(data, false);
+			} else {
+				throw ex2;
+			}
+		} catch (Exception ex3) {
+			if (this.logger.isErrorEnabled()) {
+				this.logger.error("Error sending command: " + data + " to: " + this.host + " port: " + this.port + " connection state: " + this.status.getConnectionState() + " error: ", ex3);
+			}
+
+			this.status.setLastError(ex3);
+			this.status.setConnectionState(ConnectionState.Failed);
+			this.destroyChannel();
+
+			if (retryOnError) {
+				return this.send(data, false);
+			} else {
+				throw ex3;
+			}
+		}
+	}
+
+	private String[] internalSend(String outputData) throws IOException {
+		this.write(outputData);
+		return this.read(outputData, this.socket.getInputStream());
+	}
+
+	private void write(String outputData) throws IOException {
+		if (this.socket == null) {
+			throw new IllegalStateException("Socket connection was not established. Please check target host availability and credentials.");
+		} else {
+			OutputStream os = this.socket.getOutputStream();
+			OutputStreamWriter osw = new OutputStreamWriter(os);
+			osw.write(outputData);
+			osw.flush();
+		}
+	}
+
+	private String[] read(String command, InputStream in) throws IOException {
+		if (this.logger.isDebugEnabled()) {
+			this.logger.debug("DEBUG - Socket Communicator reading after command text \"" + command + "\" was sent to host " + this.host);
+		}
+
+		// Count number of responses
+		int countResponses = 0;
+		InputStreamReader inputStreamReader = new InputStreamReader(in);
+		StringBuilder stringBuilder = new StringBuilder();
+
+		do {
+			int x = inputStreamReader.read();
+			stringBuilder.append((char) x);
+
+			if (x == 0) {
```

Comment:
Please add the comment to explain why the x == 0 here ("/00" or sth)

Reply:
sure, thank you